File: travel-dest-frontend/src/types/agent.ts
// TypeScript types for agent requests, responses, history, and tool traces.

export type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };

export type ToolCall = {
  tool_name: string;
  status: string;
  latency_ms?: number;
  tool_input?: JsonValue | Record<string, unknown> | null;
  tool_output?: JsonValue | Record<string, unknown> | null;
  output?: JsonValue | Record<string, unknown> | null;
  error_message?: string | null;
};

export type RetrievedDestination = {
  destination?: string;
  title?: string | null;
  url?: string | null;
  chunk_index?: number;
  chunk_text?: string;
  similarity_score?: number;
};

export type AgentRunRequest = {
  query: string;
};

export type AgentRunResponse = {
  agent_run_id?: string;
  answer?: string;
  final_answer?: string;
  predicted_style?: string;
  destinations?: RetrievedDestination[];
  retrieval?: {
    status?: string;
    destinations?: RetrievedDestination[];
    [key: string]: unknown;
  };
  trip_style_prediction?: {
    status?: string;
    prediction?: {
      predicted_style?: string;
      confidence?: number;
      extracted_features?: Record<string, unknown>;
      [key: string]: unknown;
    } | null;
    [key: string]: unknown;
  };
  live_conditions?: {
    status?: string;
    conditions?: Array<Record<string, unknown>>;
    [key: string]: unknown;
  };
  tool_calls?: ToolCall[];
  total_tokens?: number;
  estimated_cost?: number;
};

export type NormalizedAgentRun = {
  answer: string;
  predictedStyle?: string;
  destinations: RetrievedDestination[];
  toolCalls: ToolCall[];
  totalTokens?: number;
  estimatedCost?: number;
  raw: AgentRunResponse;
};

export type AgentRunHistoryItem = {
  id: string;
  user_query: string;
  final_answer?: string | null;
  total_tokens?: number;
  estimated_cost?: number;
  created_at?: string;
};

File: travel-dest-backend/app/schemas/agent.py
"""Schemas for the agent route, response payloads, and internal summaries."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.tools import (
    ClassifyTravelStyleOutput,
    GetLiveConditionsOutput,
    RetrieveDestinationsOutput,
)


class HealthResponse(BaseModel):
    """Small health response for quick checks."""

    status: str = "ok"


class AgentRunRequest(BaseModel):
    """Incoming user query for the smart travel planner agent."""

    query: str = Field(min_length=5)


class ToolCallLogResponse(BaseModel):
    """Tool call payload returned in agent responses and history."""

    tool_name: str
    status: str
    latency_ms: int
    tool_input: dict[str, Any] | None = None
    output: dict[str, Any]
    error_message: str | None = None


class AgentRunResponse(BaseModel):
    """Full response from one agent execution."""

    agent_run_id: UUID
    final_answer: str
    total_tokens: int
    estimated_cost: float
    retrieval: RetrieveDestinationsOutput
    trip_style_prediction: ClassifyTravelStyleOutput
    live_conditions: GetLiveConditionsOutput
    tool_calls: list[ToolCallLogResponse]


class AgentRunHistoryItem(BaseModel):
    """One persisted agent run shown in user history."""

    id: UUID
    user_query: str
    final_answer: str | None = None
    total_tokens: int
    estimated_cost: float
    created_at: datetime

    model_config = {"from_attributes": True}
